content/blog/2024-09-10-memory-access-amortization.md

```diff
@@ -0,0 +1,158 @@
+
+### Batching and interleaving 
+
+To overcome this inefficiency, we adopted the following approach. Every time a batch of incoming commands from the I/O threads is ready for execution, Valkey’s main thread efficiently prefetches the memory addresses needed for future lookupKey invocations for the keys involved in the commands  before executing the commands. This prefetch phase is achieved by dictPrefetch, which, similarly as with the linked list example from above, interleaves the table→dictEntry→...dictEntry→robj search sequences for all keys. This reduces the time spent on lookupKey by more than 80%. Another issue we had to address was that all the incoming parsed commands from the I/O threads were not present in the L1/L2 caches of the core running Valkey’s main thread. This was also resolved using the same method.  All the relevant code can be found in [memory_prefetch.c](https://github.com/valkey-io/valkey/blob/unstable/src/memory_prefetch.c). In total the impact of the memory access amortization on Valkey performance is almost 50% and it increased the requests per second to more than 1.19M rps. 
+
+
+### How to reproduce Valkey 8.0 performance numbers 
+
+This section will walk you through the process of reproducing our performance results, where we achieved 1.19 million requests per second using Valkey 8.
+
+### Hardware Setup
+
+We conducted our tests on an AWS EC2 c7g.4xlarge instance, featuring 16 cores on an ARM-based (aarch64) architecture.
+
+### System Configuration
+
+> Note: The core assignments used in this guide are examples. Optimal core selection may vary depending on your specific system configuration and workload.
+
+Interrupt affinity - locate the network interface with `ifconfig` (let's assume it is `eth0`) and its associated IRQs with 
+<pre class="inverted-code">
+grep eth0 /proc/interrupts | awk '{print $1}' | cut -d : -f 1
+</pre>
+In our setup, lines `48` to `55` are allocated for `eth0` interrupts. Allocate one core per 4 IRQ lines: 
+<pre class="inverted-code">
+for i in {48..50}; do echo 1000 > /proc/irq/$i/smp_affinity; done
+for i in {51..55}; do echo 2000 > /proc/irq/$i/smp_affinity; done  
+</pre>
+Server configuration - launch the Valkey server with these minimal configurations:
+<pre class="inverted-code">
+./valkey-server --io-threads 9 --save --protected-mode no
+</pre>
+`--save` disables dumping to RDB file and `--protected-mode no `  allows connections from external hosts. `--io-threads` number includes the main thread and the IO threads, meaning that in our case 8 I/O threads are launched in addition to the main thread. 
+
+Main thread affinity - pin the main thread to a specific CPU core, avoiding the cores handling IRQs. Here we use core #3:
+<pre class="inverted-code">
+sudo taskset -cp 3 `pidof valkey-server`
+</pre>
+> Important: We suggest to experiment with different other to find the optimal performance while avoiding conflicts with the IRQ-handling cores.
```

Comment:
`We suggest to experiment with different other to find...` doesn't quite parse. Something is missing.

Do you mean something like `We suggest to experiment with different cores to find...`?

Reply:
We suggest experimenting with different core pinning strategies to find the optimal performance while avoiding conflicts with IRQ-handling cores